Adding developer doc  for proxy create by upload OAS flow (#1788)
wso2/docs-choreo-dev: en/developer-docs/docs/develop-components/develop-proxy/develop-an-api-proxy-from-an-open-api-specification.md

# Develop an API Proxy from Open API Specification

> An API proxy acts as an intermediary between an existing API and Choreo, intercepting all requests made to the API. It also functions as a managed API, allowing you to apply essential API management features such as security policies and rate limiting.

Reviewer: should we do a mention about the fact that v2 and v3 is supported.

Author: What if it mention under prerequisites section ?

Reviewer: I guess it is okay to ignore this given that we gonna support 3.1 soon

Author: we can ignore this as the limitation (oas version compatibility) will resolve soon.

In this guide, you will:

- Create an API proxy component to expose an existing API.
- Deploy the API proxy.
- Test the API proxy to verify its functionality.
- Manage the API.
- Consume the API.

## Prerequisites

- If you're signing in to the Choreo Console for the first time, create an organization:
    1. Go to [https://console.choreo.dev/](https://console.choreo.dev/) and sign in using your preferred method.
    2. Enter a unique organization name. For example, `Stark Industries`.
    3. Read and accept the privacy policy and terms of use.
    4. Click **Create**.

    This creates the organization and opens the **Project Home** page of the default project created for you.

## Step 1: Create an API proxy

> To create an API proxy, you can either upload an OpenAPI specification or provide an OpenAPI specification URL. In this guide, you will specify a URL to an OpenAPI definition of a sample API.

Reviewer: ```suggestion
To create an API proxy, you can either upload an OpenAPI document or provide a hosted OpenAPI document's URL. In this guide, you will specify a URL to an OpenAPI definition of a sample API.
```

> To create an API proxy, you can either upload an OpenAPI specification or provide an OpenAPI specification URL. In this guide, you will specify a URL to an OpenAPI definition of a sample API.

Reviewer: ```suggestion
To create an API proxy, you can either upload an OpenAPI document or provide a hosted OpenAPI document's URL. In this guide, you will specify a URL to an OpenAPI definition of a sample API.
```
https://swagger.io/specification/ openapi document seemes to be correct term.

1. Go to [https://console.choreo.dev/](https://console.choreo.dev/) and sign in. This opens the project home page.
2. If you already have one or more components in your project, click **+ Create**. Otherwise, proceed to the next step.
3. Click the **API Proxy** card. This opens the **Create an API Proxy** pane.
4. Select **Upload API Specification** option. 

> 5. Click on **Try with sample URL** to go ahead with the sample Open API Specification.

Reviewer: ```suggestion
5. Click on **Try with sample URL** to go ahead with the sample Open API Document.
```

>     !!! info

Reviewer: I think we can get rid of this. This is more like a tutorial anyway. So better to stick to our sample and move on.

        To streamline the process, if you already have a specification file saved on your device, simply upload it directly instead of relying on the sample.

6. Specify the following values as Proxy Metadata:

    !!! note

>         Values will auto populate according to the data in the uploaded/selected specification file. Following values are for the sample provided by choreo.

Reviewer 2: Suggestion based on the next comment:

While most values are auto-generated from the specification, the context value has been explicitly set to `petstore` to enhance clarity

    | **Field**       | **Value**                                  |
    |-----------------|--------------------------------------------|

>     | **Context**     | `api/v3`                                   |

Reviewer 2: Shall we ask the user to change the context as the default one is a bit untidy? (it doesn't contain the API name and it has an additional version part as well)
I know its a little bit imperfect to ask for changing, but I feel its not good to move forward with the default value. suggestion: `petstore`
We might need to slightly alter the previous sentence as well, as we are changing the context - `Following values are for the sample provided by choreo.`

Reviewer 2: https://github.com/wso2/docs-choreo-dev/pull/1788/files#r2006992742

Reviewer 2: Update the screenshot as well accordingly.

    | **Version**     | `v1.0`                                      |
    | **Target**      | `https://petstore3.swagger.io/api/v3` |

7. Enter following details for the Component Details:

    !!! info
        The **Component Name** field must be unique and cannot be changed after creation.

    | **Field**                 | **Value**          |
    |---------------------------|--------------------|

>     | **Component Display Name**| `Pet Store Service`          |

Reviewer 2: ```suggestion
    | **Component Display Name**| `Petstore Service`          |
```

    | **Component Name**        | `pet-store-service`          |

>     | **Description**           | `Sample proxy for pet store service`     |

Reviewer 2: ```suggestion
    | **Description**           | `Sample proxy for Petstore service`     |
```

8. Click **Create**. This creates the API proxy component and takes you to the **Deploy** page.

## Step 2: Deploy the API proxy

> 1. In the left navigation menu, click **Deploy**.

Reviewer: Should we have statement saying that you must deploy it in order to invoke it.

Author: No point of that instruction, right. Since choreo console handle it already (console has info message). 
That is not a blocker at all even the user doesn't follow each steps of the document.

Now you are ready to test the API proxy.

## Step 3: Test the API proxy

Choreo allows you to test your API proxy using either the [integrated OpenAPI Console](../../testing/test-rest-endpoints-via-the-openapi-console.md) or [cURL](../../testing/test-apis-with-curl.md). In this guide, you will use the OpenAPI Console.

!!! tip
    Choreo enables OAuth 2.0 to secure APIs by default. Therefore, you need an access token to invoke an API.

    - Choreo automatically generates a key to test the API via the OpenAPI Console. To view the key, click the show key icon in the **Security Header** field.
    - To disable security for the entire API or a specific resource:
        1. In the left navigation menu, click **Deploy**.
        2. Go to the **Build Area** card and click **Security Settings**.
        3. In the **Security Settings** pane:
            - To disable security for the entire API, clear the **OAuth2** checkbox.
            - To disable security for a specific resource, expand the relevant resource and turn off the **Security** toggle.

>         4. Click **Apply**.

Reviewer: I think unless we deploy it, the changes won't apply.

Reviewer 2: Yup.. we have to deploy..
I think this has to be fixed in the other two documents (from scratch and git proxy) too

Author: Added that point next to the 4th  point.

4. Select the status **available** from the dropdown and click **Execute**. You will see a response similar to the following:

    ![API proxy response](../../assets/img/develop-components/develop-a-rest-api-proxy/try-out-pet-store-response.png){.cInlineImage-full}

    This indicates that your API proxy is working as expected.

## Step 4: Manage the API proxy

Now that you have a tested API proxy, you can publish it and make it available for application developers to consume. In this guide, you will apply rate limiting to the API and publish it.

### Step 4.1: Apply rate limiting to the API proxy

1. In the left navigation menu, click **Deploy**.
2. Go to the required environment card and click the settings icon corresponding to **CORS, Rate Limiting and Resiliency**.
3. In the **CORS, Rate Limiting and Resiliency** pane, click **Rate Limiting** to expand the section.
4. Select **API Level** as the **Rate Limiting Level**.
5. Specify appropriate values for the **Request Limit** and **Time Unit** fields. You can proceed with the default values.
6. Click **Apply**. This applies the rate limiting level to the API proxy and redeploys it.

> ### Step 4.2: Publish the API proxy

Reviewer: Let's do a mention about the fact that, For API to appear in Developer Portal you should publish the API.

Author: ![Screenshot 2025-03-25 at 15 39 57](https://github.com/user-attachments/assets/3017b417-246f-498e-86ba-00de73aae307)

Reviewer: I mean emphasizing the reason why you need to publish an API. Typical user won't get this neh

Author: got it.
what about this 

![Screenshot 2025-03-27 at 15 36 24](https://github.com/user-attachments/assets/446b732c-f1b9-463d-9336-c05055cf8d2e)

1. In the left navigation menu, click **Lifecycle** under **Manage**. This takes you to the **Lifecycle** page.
2. Click **Publish**.
3. In the **Publish API** dialog, click **Confirm** to proceed with publishing the API. If you want to change the display name, make the necessary changes and then click **Confirm**. This changes the API lifecycle state to **Published**.

> ## Step 5: Invoke the API

Reviewer: ```suggestion
## Step 5: Invoke the API From DevPortal
```

To generate credentials for the published API and invoke it via the Choreo Developer Portal, follow these steps:

> 1. In the **Lifecycle** page, click **Go to Devportal**. This takes you to the `Pet Store Service` in the Choreo Developer Portal.

Reviewer 2: ```suggestion
1. In the **Lifecycle** page, click **Go to Devportal**. This takes you to the `Petstore Service` in the Choreo Developer Portal.
```

> 2. **Invoke the API**:

Reviewer: ```suggestion
2. Invoke the API:
```
